src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -357,63 +365,171 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
   }
 
   def internalUpdateCluster(existingCluster: Cluster, clusterRequest: ClusterRequest) = {
-    if(existingCluster.status.isUpdatable) {
+    implicit val booleanSumMonoidInstance = new Monoid[Boolean] {
+      def empty = false
+      def combine(a: Boolean, b: Boolean) = a || b
+    }
+
+    if (existingCluster.status.isUpdatable) {
       for {
-        _ <- maybeUpdateAutopauseThreshold(existingCluster.id, clusterRequest.autopause, clusterRequest.autopauseThreshold)
+        autopauseChanged <- maybeUpdateAutopauseThreshold(existingCluster, clusterRequest.autopause, clusterRequest.autopauseThreshold).attempt
 
-        clusterResized <- maybeResizeCluster(existingCluster, clusterRequest.machineConfig)
+        clusterResized <- maybeResizeCluster(existingCluster, clusterRequest.machineConfig).attempt
 
-        // Set the cluster status to Updating only if the cluster was resized
-        _ <- if(clusterResized) { dbRef.inTransaction { _.clusterQuery.updateClusterStatus(existingCluster.id, ClusterStatus.Updating) } } else Future.successful(0)
+        masterMachineTypeChanged <- maybeChangeMasterMachineType(existingCluster, clusterRequest.machineConfig).attempt
+
+        masterDiskSizeChanged <- maybeChangeMasterDiskSize(existingCluster, clusterRequest.machineConfig).attempt
+
+        // Note: only resizing a cluster triggers a status transition to Updating
+        (errors, shouldUpdate) = List(
+          autopauseChanged.map(_ => false),
+          clusterResized,
+          masterMachineTypeChanged.map(_ => false),
+          masterDiskSizeChanged.map(_ => false)
+        ).separate
+
+        // Set the cluster status to Updating if the cluster was resized
+        _ <- if (shouldUpdate.combineAll) {
+          dbRef.inTransaction { _.clusterQuery.updateClusterStatus(existingCluster.id, ClusterStatus.Updating) }.void
+        } else Future.successful(())
+
+        cluster <- errors match {
+          case Nil => internalGetActiveClusterDetails(existingCluster.googleProject, existingCluster.clusterName)
+          // Just return the first error; we don't have a great mechanism to return all errors
```

Comment:
Yeah, in the past we intentionally tried to _not_ use `ErrorReport` to avoid leaking internal details like stack traces to the user. But something similar might serve this purpose of returning multiple validation errors.